circleguard/investigator.py

```diff
@@ -206,11 +206,153 @@ def median_frametime(replay):
+                        filtered_frametimes_more.extend([filtered_frametimes[i-1], filtered_frametimes[i]])
+                    else:
+                        filtered_frametimes_more.append(combined_time)
+                else:
+                    if np.abs(median - filtered_frametimes[i]) > 4:
+                        combine_next = True
+                    else:
+                        combine_next = False
+                        filtered_frametimes_more.append(filtered_frametimes[i])
+            filtered_frametimes = filtered_frametimes_more
+        
+        # use different calculation if player is using low refresh rate keyboard
+        missing_numbers = 0
+        for i in range(max(1, min(unique)), int(np.median(filtered_frametimes_more))):
+            if i not in unique:
+                missing_numbers += 1
+        if missing_numbers > 2:
+            lowhz_keyboard = True
+        else:
+            lowhz_keyboard = False
+
+        return np.median(filtered_frametimes_more), np.array(filtered_frametimes_more), lowhz_keyboard
+
+    @staticmethod
+    def average_frametime(replay):
+        """
+        Attempt at a more precise and accurate average frametime.
+
+        avg frametime < 16 is slightly suspicious, check other replays from player
+        avg frametime < 15.5 is very suspicious, 90-99% certainty, manually check replay and check other replays from player
+        avg frametime < 15 is ~99.95% sure it's timewarp
+        avg frametime < 14 is ~99.999% sure it's timewarp
+        avg frametime < 13 is 100% timewarp
+        """
+
+        median, frametimes, lowhz_keyboard = Investigator._filter_frametime(replay)
+        unique, counts = np.unique(frametimes, return_counts=True)
+
+        # exception for 100hz polling keyboards
+        if lowhz_keyboard:
+            # removes some lower frametimes that were likely missed during filtering
+            counts_cutoff = max(counts)/6
+            remove_index = []
+            for i in range(len(counts)):
+                if unique[i] < median - 2 and counts[i] < counts_cutoff:
+                    remove_index.append(i)
+            counts = np.delete(counts, remove_index)
+            unique = np.delete(unique, remove_index)
+            filtered_frametimes = frametimes[np.isin(frametimes, unique)]
+            # increase by a bit since replays with low fps tend to be a bit lower (due to shorter frametimes being a bit harder to accurately filter out)
+            return np.average(filtered_frametimes) + 0.1
+        else:
+            # filters out low frametimes and grabs frametimes around the peaks
+            counts_cutoff = max(counts)/4
+            frametime_cutoff = 6
+            number_a = 7
+            if replay.mods.value & 1 << 6:
+                frametime_cutoff *= 1.5
+                number_a *= 1.5
+            if replay.mods.value & 1 << 8:
+                frametime_cutoff *= 0.75
+                number_a *= 0.75
```

Comment:
should be using `replay.mods & Mod.DT` instead of this low level manipulation, but also this should use `utils.convert_statistic` which I believe is exactly this calculation